### DISCUSSION TITLE
Tutorial for integrating Caddy v2 and PocketBase into a single app

### DISCUSSION THREAD

**Discussion by W1M0R** [pocketbase/pocketbase#2353] [Q&A]
Reading through some of the existing issues and feature requests, many of the suggestions seem to fall into the realm of things that are solved by Caddy and Nginx. 

Instead of re-inventing the wheel for many of these issues, maybe it would be worthwhile to expend the effort to create a guide showing how to integrate Caddy and PocketBase into a single app, and then using Caddy's rate limiting (and other functionality) to implement the things that fall outside the realm of PocketBase. The user still gets a single binary in the end, but with maximum flexibility.

I wish I could offer to do this, but unfortunately I have no experience doing this (yet), and have only begun researching the possibility.

Some related reading:

1. https://medium.com/backendarmy/embedding-a-web-server-in-go-e116cbbdc936
2. https://github.com/caddyserver/xcaddy
3. https://caddyserver.com/docs/extending-caddy
4. https://caddyserver.com/docs/architecture
5. https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MTM2Nw==#apps-457a9a42-afca-4f05-a867-762a6d29e244
6. https://caddy.community/t/how-can-i-embed-caddy-2-within-my-application/10021/2
7. https://caddy.community/t/app-module-example-e-g-implement-a-simple-blogging-system/12391
8. https://github.com/Rex--/caddy-git-server/blob/a7bad1eb558bd85d0ce932c972c43a51805a20f9/git_server.go#L161
9. https://caddy.community/t/example-of-go-embed-in-v2/6683

**Comment by mukhtharcm** [comment]
Did you ended up doing something?

**Reply by W1M0R** [reply]
No I didn't.

**Reply by dschissler** [reply]
You would need to take the code out of [serve.go](https://github.com/pocketbase/pocketbase/blob/master/apis/serve.go) to integrate it as a Caddy module (or whatever term that they use for that).  Just getting something barely hacked together might not be that complicated.  It's probably not very viable right now, but you'd learn a lot, and maybe you'd be able to express some limitations that ganigeorgiev could work making more flexible.

Pocketbase is currently using Echo for the router.  ganigeorgiev has stated that the next major refactor (that is maybe due soon or in many months from now) is going to a standard (r, w) Golang http handler.  So if you hit a wall with that then I recommend not wasting time on it until the refactor is released to us.  ganigeorgiev has been very cagey and grumpy about informing us, and so you'll be unable to get any information on that at all, and he's told us outright that he doesn't want us asking.  So if Echo becomes an issue for you then I recommend just putting the effort on pause because you'd be just throwing any workarounds and glue code away.

**Comment by mlusetti** [comment]
There's a caddy module recently added.
You can find it here: https://github.com/mohammed90/caddy-pocketbase
Here is the docs: https://caddyserver.com/docs/json/apps/pocketbase/
